### PR TITLE
Expansion saved, very basic event system: REVIEW

### PR DESCRIPTION
**Pull request by ThompsonTyler**
@flo Wanted a review PR for currently progress and for feedback regarding event system.

Currently to track events I'm watching any possible event that is considered an event(this list will increase as more events are added) and then when one of those occurs the root entity will check it's list of entities and see if that event matches any of them, if so it sends a EventTriggerEvent(holds any relevant data needed for triggers later down the line) to that event entity. The event then processes it(will eventually check against conditionals) and then sends it to it's actions. Actions if they receive and event will just trigger and do what they are supposed to do.

This works and might be fine, but in the end it basically will have to track everything even if they are not used and will still check every eventEntity for every event that occurs, this results in terrible runtime with large numbers and can definitely be optimized. Not sure if there is a way to subscribe/unsubscribe the @ReceiveEvent annotation so that it can be turned on/off during runtime as needed. This reduces some un-necessary monitoring, but still doesn't solve the checking each event's "condition/trigger" one at a time.

Expansion also works here, but there shouldn't be too much feedback on that, it works, feedback would be more on the actual code and methods in order to accomplish expansion or better solutions vs needing solutions in the first place.

### CODE REVIEW

**Comment by ThompsonTyler** [comment]
PR is now meant for actual review and pull, with commit cf2c3d9

File: src/main/java/org/terasology/scenario/components/ScenarioComponent.java [hunk 1]
```diff
@@ -18,11 +18,12 @@
...
-import java.util.List;
+import java.util.HashSet;
+import java.util.Set;
 
 /**
  * Component attached to a scenario entity. Denotes the "root" of the logic structure.
  */
 public class ScenarioComponent implements Component {
-    public List<EntityRef> triggerEntities;
+    public Set<EntityRef> triggerEntities  = new HashSet<>();
```
Comment (flo) [inline src/main/java/org/terasology/scenario/components/ScenarioComponent.java]: Why a set, what about the order?

File: src/main/java/org/terasology/scenario/components/TriggerActionListComponent.java [hunk 2]
```diff
@@ -18,11 +18,12 @@
...
+import java.util.Set;
 
 /**
- * Component attached to event entities that allows for a list of actions to be attached
+ * List for actions in a trigger
  */
-public class ActionListComponent implements Component {
-    public List<EntityRef> actions;
+public class TriggerActionListComponent implements Component {
+    public Set<EntityRef> actions = new HashSet<>();
```
Comment (flo) [inline src/main/java/org/terasology/scenario/components/TriggerActionListComponent.java]: Especually for actions I think an order could be very important.

E.g. imagine the actions
1. Create item of type brick
2. Set item stack size of last created item to 5

File: src/main/java/org/terasology/scenario/components/TriggerConditionListComponent.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+import org.terasology.entitySystem.entity.EntityRef;
+
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * List for conditions in a trigger
+ */
+public class TriggerConditionListComponent implements Component {
+    public Set<EntityRef> conditions = new HashSet<>();
```
Comment (flo) [inline src/main/java/org/terasology/scenario/components/TriggerConditionListComponent.java]: Same here, I think a list might be more important

File: src/main/java/org/terasology/scenario/components/TriggerEventListComponent.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+import org.terasology.entitySystem.entity.EntityRef;
+
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * List for events in a trigger
+ */
+public class TriggerEventListComponent implements Component {
+    public Set<EntityRef> events = new HashSet<>();
```
Comment (flo) [inline src/main/java/org/terasology/scenario/components/TriggerEventListComponent.java]: Even events we mightcwant to allow the user to give them a meaningfull order

File: src/main/java/org/terasology/scenario/components/TriggerNameComponent.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+/**
+ * Component that includes the details about a trigger, currently it is a name and three empty entities that are
+ * used for the entity expansion of a hub tool that represent the blank event/condition/action portions of the UI
+ */
+public class TriggerNameComponent implements Component {
+    public String name;
+
+    public EntityRef entityForEvent;
+    public EntityRef entityForCondition;
+    public EntityRef entityForAction;
```
Comment (flo) [inline src/main/java/org/terasology/scenario/components/TriggerNameComponent.java:30]: I think you arr aiming to have for example entityForAction contain the action list.
Soif the Ui displays

MyTrigger
- events (entityForEvent)
  - player respawns
- conditions (entityForCindition)
- actions (entityForAction)
  - Do X
  - Do Y


I think however that we should not have this 3 entities. I think they could be always expanded. So we can have the nodes without having entities for them. So the trigger entity would have 4  nodes. Thw node that represents the trigger and the 3 nodes that represents the evsnts, cinditions and actions. 

That makes not just the entity structure simpler but also is more future save.

But it is a point we can discuss

File: src/main/java/org/terasology/scenario/internal/systems/ActionEventSystem.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    @Override
+    public void initialise() {
+        super.initialise();
+        blockItemFactory = new BlockItemFactory(entityManager);
+    }
+
+    @ReceiveEvent
+    public void onEventTriggerEvent(EventTriggerEvent event, EntityRef entity, ActionComponent action) {
+        switch (action.type) {
+            case GIVE_ITEM:
```
Comment (flo) [inline src/main/java/org/terasology/scenario/internal/systems/ActionEventSystem.java:54]: As mentionend before I think the enum should be removed. Instead the specific action prefab should have a component thatis unique to the action and that can be added to eother pararameter list or ReceiveEvent annotation values to have the handler only run for actions of that type

**Comment by flo** [comment]
There is a small compile error in ScenarioRootManagementSystem from a import of ActionListComponent which no longer exist.

Also a minor thing which not need to be fixed for this PR: There seems to be no limitation in which way nodes can be moved around. Thus it is possible to is possible to move nodes around in a way that does not make sense.

**Comment by flo** [comment]
I noticed that there are still some movement bugs (the events,actions and conditions can be moved around within their level and they vanish when you drop them.) but I think we can ignore that for now.

**Comment by flo** [comment]
I think it is more important to get more of the other fundamentals down like the parameter editing.

**Pull request merged**